// @module-federation/nextjs-mf/src/plugins/NextFederationPlugin.js
/*
    MIT License http://www.opensource.org/licenses/mit-license.php
    Author Zackary Jackson @ScriptedAlchemy
*/
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.NextFederationPlugin = void 0;
const tslib_1 = require("tslib");
const utilities_1 = require("@module-federation/utilities");
const path_1 = tslib_1.__importDefault(require("path"));
const internal_1 = require("../internal");
const AddRuntimeRequirementToPromiseExternalPlugin_1 = tslib_1.__importDefault(require("./AddRuntimeRequirementToPromiseExternalPlugin"));
const ChildFederationPlugin_1 = tslib_1.__importDefault(require("./ChildFederationPlugin"));
const DevHmrFixInvalidPongPlugin_1 = tslib_1.__importDefault(require("./DevHmrFixInvalidPongPlugin"));
class NextFederationPlugin {
    constructor(options) {
        const { extraOptions, ...mainOpts } = options;
        this._options = mainOpts;
        this._extraOptions = {
            automaticPageStitching: false,
            enableImageLoaderFix: false,
            enableUrlLoaderFix: false,
            skipSharingNextInternals: false,
            automaticAsyncBoundary: false,
            ...extraOptions,
        };
    }
    apply(compiler) {
        if (!compiler.options.name) {
            throw new Error('name is not defined in Compiler options');
        }
        if (!this._options.filename) {
            throw new Error('filename is not defined in NextFederation options');
        }
        if (!['server', 'client'].includes(compiler.options.name)) {
            return;
        }
        const isServer = compiler.options.name === 'server';
        const { webpack } = compiler;
        if (isServer) {
            // target false because we use our own target for node env
            compiler.options.target = false;
            const { StreamingTargetPlugin } = require('@module-federation/node');
            new StreamingTargetPlugin(this._options, {
                ModuleFederationPlugin: webpack.container.ModuleFederationPlugin,
            }).apply(compiler);
            this._options.library = {
                type: 'commonjs-module',
                name: this._options.name,
            };
            // output remote to ssr if server
            this._options.filename = this._options.filename.replace('/chunks', '/ssr');
            // should this be a plugin that we apply to the compiler?
            (0, internal_1.internalizeSharedPackages)(this._options, compiler);
        }
        else {
            if (this._extraOptions.automaticPageStitching) {
                compiler.options.module.rules.push({
                    test: /next[\\/]dist[\\/]client[\\/]page-loader\.js$/,
                    loader: path_1.default.resolve(__dirname, '../loaders/patchNextClientPageLoader'),
                });
            }
            if (this._options.remotes) {
                this._options.remotes = (0, internal_1.parseRemotes)(this._options.remotes);
            }
            if (this._options.library) {
                console.error('[mf] you cannot set custom library');
            }
            this._options.library = {
                // assign remote name to object to avoid SWC mangling top level variable
                type: 'window',
                name: this._options.name,
            };
        }
        //patch next
        compiler.options.module.rules.push({
            test(req) {
                if (req.includes(path_1.default.join(compiler.context, 'pages')) || req.includes(path_1.default.join(compiler.context, 'app'))) {
                    return /\.(js|jsx|ts|tsx|md|mdx|mjs)$/i.test(req);
                }
                return false;
            },
            include: compiler.context,
            exclude: /node_modules/,
            loader: path_1.default.resolve(__dirname, '../loaders/patchDefaultSharedLoader'),
        });
        if (this._extraOptions.automaticAsyncBoundary) {
            compiler.options.module.rules.push({
                test: (request) => {
                    if (request.includes(path_1.default.join(compiler.context, 'pages')) || request.includes(path_1.default.join(compiler.context, 'app'))) {
                        return /\.(js|jsx|ts|tsx|md|mdx|mjs)$/i.test(request);
                    }
                    return false;
                },
                exclude: [/node_modules/, /_document/, /_middleware/, /pages\/api/],
                resourceQuery: (query) => !query.includes('hasBoundary'),
                loader: path_1.default.resolve(__dirname, '../loaders/async-boundary-loader')
            });
        }
        //todo runtime variable creation needs to be applied for server as well. this is just for client
        // TODO: this needs to be refactored into something more comprehensive. this is just a quick fix
        new webpack.DefinePlugin({
            'process.env.REMOTES': (0, utilities_1.createRuntimeVariables)(this._options.remotes),
            'process.env.CURRENT_HOST': JSON.stringify(this._options.name),
        }).apply(compiler);
        const ModuleFederationPlugin = isServer
            ? require('@module-federation/node').NodeFederationPlugin
            : webpack.container.ModuleFederationPlugin;
        // ignore edge runtime and middleware builds
        if (!ModuleFederationPlugin) {
            return;
        }
        const internalShare = (0, internal_1.reKeyHostShared)(this._options.shared);
        const hostFederationPluginOptions = {
            ...this._options,
            exposes: {},
            shared: {
                noop: {
                    import: 'data:text/javascript,module.exports = {};',
                    requiredVersion: false,
                    eager: true,
                    version: '0',
                },
                ...internalShare,
            },
        };
        new ModuleFederationPlugin(hostFederationPluginOptions, {
            ModuleFederationPlugin,
        }).apply(compiler);
        new ChildFederationPlugin_1.default(this._options, this._extraOptions).apply(compiler);
        new AddRuntimeRequirementToPromiseExternalPlugin_1.default().apply(compiler);
        if (compiler.options.mode === 'development') {
            new DevHmrFixInvalidPongPlugin_1.default().apply(compiler);
        }
    }
}
exports.NextFederationPlugin = NextFederationPlugin;
exports.default = NextFederationPlugin;
